File: lib/winrmRun.js
const { spawn } = require('child_process');
const path = require('path');

const SCRIPT_PATH = path.join(__dirname, 'winrm_run.py');
const TIMEOUT_MS = 20000; // matches winrmSpecs.js - long enough for a real query, short enough not to hang a modal indefinitely on an unreachable host

// General-purpose counterpart to fetchWinrmSpecs - runs whatever
// PowerShell script is passed in, rather than a single hardcoded query.
// Credentials go to the Python helper over stdin, not argv, for the same
// reason as winrmSpecs.js: argv is visible to other processes on the same
// host via `ps aux` / /proc/<pid>/cmdline, stdin isn't.
function runWinrmPowerShell({ hostname, port, username, password, useSsl, powershell, timeoutMs }) {
  return new Promise((resolve) => {
    const child = spawn('python3', [SCRIPT_PATH]);
    let stdout = '';
    let stderr = '';
    let settled = false;

    const finish = (result) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve(result);
    };

    const timer = setTimeout(() => {
      child.kill('SIGKILL');
      finish({ error: 'Timed out waiting for a response from the remote machine.' });
    }, timeoutMs || TIMEOUT_MS);

    child.stdout.on('data', (chunk) => { stdout += chunk; });
    child.stderr.on('data', (chunk) => { stderr += chunk; });

    child.on('error', (err) => {
      finish({ error: `Could not start the WinRM helper: ${err.message}` });
    });

    child.on('close', () => {
      if (settled) return;
      try {
        const parsed = JSON.parse(stdout.trim());
        finish(parsed);
      } catch (e) {
        finish({ error: stderr.trim() || 'The WinRM helper produced no usable output.' });
      }
    });

    child.stdin.write(JSON.stringify({ hostname, port, username, password, use_ssl: Boolean(useSsl), powershell }));
    child.stdin.end();
  });
}

module.exports = { runWinrmPowerShell };

File: lib/winrm_run.py
#!/usr/bin/env python3
"""
Runs an arbitrary PowerShell script on a Windows machine over WinRM and
prints its JSON output.

Invoked by lib/winrmRun.js as a subprocess. Reads a single JSON object from
stdin: {"hostname": ..., "port": ..., "username": ..., "password": ...,
"use_ssl": bool, "powershell": "<script text>"}. Credentials travel over
stdin rather than as command-line arguments, same reasoning as
winrm_specs.py - argv is visible to other processes on the same host via
`ps aux` / /proc/<pid>/cmdline, stdin isn't.

Always exits 0 and prints exactly one JSON object to stdout - either the
script's parsed output, or {"error": "..."} - so the Node side has one
simple, uniform way to parse the result.

This is the general-purpose counterpart to winrm_specs.py, which stays as
its own separate, already-tested script rather than being rewritten to
call through this one - lower risk than refactoring something that's
already confirmed working end-to-end against a real machine.
"""
import sys
import json
import winrm
from winrm.exceptions import WinRMTransportError, InvalidCredentialsError


def main():
    try:
        params = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        print(json.dumps({"error": f"Internal error reading request: {e}"}))
        return

    hostname = params.get("hostname")
    port = params.get("port", 5985)
    username = params.get("username")
    password = params.get("password")
    use_ssl = params.get("use_ssl", False)
    powershell = params.get("powershell", "")

    scheme = "https" if use_ssl else "http"
    endpoint = f"{scheme}://{hostname}:{port}/wsman"

    try:
        session = winrm.Session(endpoint, auth=(username, password), transport="ntlm", server_cert_validation="ignore")
        result = session.run_ps(powershell)
    except InvalidCredentialsError:
        print(json.dumps({"error": "Authentication failed - check the username and password."}))
        return
    except WinRMTransportError as e:
        print(json.dumps({"error": f"Could not reach WinRM on {hostname}:{port} - {e}"}))
        return
    except Exception as e:
        print(json.dumps({"error": f"Could not reach WinRM on {hostname}:{port} - {e}"}))
        return

    if result.status_code != 0:
        stderr_text = result.std_err.decode("utf-8", errors="replace").strip()
        print(json.dumps({"error": stderr_text or "The remote command failed with no error output."}))
        return

    stdout_text = result.std_out.decode("utf-8", errors="replace").strip()
    if not stdout_text:
        # A valid, empty result (e.g. Get-Service returning zero rows via
        # @(...) | ConvertTo-Json) still prints "[]" - genuinely empty
        # stdout means the command produced no output at all, worth
        # distinguishing from an actual empty list.
        print(json.dumps({"error": "The remote command produced no output."}))
        return
    try:
        parsed = json.loads(stdout_text)
    except json.JSONDecodeError:
        print(json.dumps({"error": "Got a response, but couldn't parse it as JSON."}))
        return

    print(json.dumps(parsed))


if __name__ == "__main__":
    main()
